File: Chatmvp/app/src/main/java/nosql/anew/chatmvp/presenters/ChatPresenter.java
package nosql.anew.chatmvp.presenters;

import com.google.firebase.database.DataSnapshot;

import java.util.ArrayList;

import nosql.anew.chatmvp.interfaces.FirebaseCallBacks;
import nosql.anew.chatmvp.interfaces.ModelCallBacks;
import nosql.anew.chatmvp.utils.FirebaseManager;
import nosql.anew.chatmvp.models.ChatPojo;
import nosql.anew.chatmvp.models.MessageModel;
import nosql.anew.chatmvp.views.IChatView;

/**
 * Created by saksham on 26/6/17.
 */
public class ChatPresenter implements FirebaseCallBacks,ModelCallBacks {

    private IChatView mIChatView;
    private MessageModel mModel;

    public ChatPresenter(IChatView iChatView){
        this.mIChatView=iChatView;
        this.mModel=new MessageModel();
    }

    public void sendMessageToFirebase(String roomName, String message) {
        if (!message.trim().equals("")){
            FirebaseManager.getInstance(roomName,this).sendMessageToFirebase(message);
        }
        mIChatView.clearEditText();
    }

    public void setListener(String roomName) {
        FirebaseManager.getInstance(roomName,this).addMessageListeners();
    }

    public void onDestory(String roomName) {
        FirebaseManager.getInstance(roomName,this).removeListeners();
        FirebaseManager.getInstance(roomName,this).destroy();
        mIChatView=null;
    }

    @Override
    public void onNewMessage(DataSnapshot dataSnapshot) {
        mModel.addMessages(dataSnapshot,this);
    }

    @Override
    public void onModelUpdated(ArrayList<ChatPojo> messages) {
        if (messages.size()>0) {
            mIChatView.updateList(messages);
        }
    }
}

File: Chatmvp - Kotlin/app/src/main/java/nosql/anew/kotlinmvp/presenters/ChatPresenter.kt
package nosql.anew.kotlinmvp.presenters

import com.google.firebase.database.DataSnapshot
import nosql.anew.kotlinmvc.interfaces.FirebaseCallBacks
import nosql.anew.kotlinmvc.interfaces.ModelCallBacks
import nosql.anew.kotlinmvp.models.ChatPojo
import nosql.anew.kotlinmvp.models.MessageModel
import nosql.anew.kotlinmvp.utils.FirebaseManager
import nosql.anew.kotlinmvp.views.IChatView
import java.util.ArrayList

/**
 * Created by Sam on 27/04/18.
 */
class ChatPresenter(private var mIChatView: IChatView?) : FirebaseCallBacks, ModelCallBacks {
    private val mModel: MessageModel

    init {
        this.mModel = MessageModel(this)
    }

    fun sendMessageToFirebase(roomName: String?, message: String) {
        if (!message.trim().equals("")) {
            FirebaseManager.getInstance(roomName!!, this)!!.sendMessageToFirebase(message)
        }
        mIChatView!!.clearEditText()
    }

    fun setListener(roomName : String?) {
        FirebaseManager.getInstance(roomName!!, this)!!.addMessageListeners()
    }

    fun onDestroy(roomName: String?) {
        FirebaseManager.getInstance(roomName!!, this)!!.removeListeners()
        FirebaseManager.getInstance(roomName, this)!!.destroy()
        mIChatView = null
    }

    override fun onNewMessage(dataSnapshot: DataSnapshot?) {
        mModel.addMessages(dataSnapshot!!)
    }

    override fun onModelUpdated(messages: ArrayList<ChatPojo>) {
        if (messages.size > 0) {
            mIChatView!!.updateList(messages)
        }
    }
}
